docarray/array/array/sequence_indexing_mixin.py

```diff
@@ -0,0 +1,216 @@
+import abc
+from typing import (
+    Any,
+    Iterable,
+    MutableSequence,
+    Optional,
+    Sequence,
+    TypeVar,
+    Union,
+    cast,
+    no_type_check,
+    overload,
+)
+
+import numpy as np
+
+from docarray.utils.misc import is_torch_available
+
+T_item = TypeVar('T_item')
+T = TypeVar('T', bound='IndexingSequenceMixin')
+
+IndexIterType = Union[slice, Iterable[int], Iterable[bool], None]
+
+
+def _is_np_int(item: Any) -> bool:
+    dtype = getattr(item, 'dtype', None)
+    ndim = getattr(item, 'ndim', None)
+    if dtype is not None and ndim is not None:
+        try:
+            return ndim == 0 and np.issubdtype(dtype, np.integer)
+        except TypeError:
+            return False
+    return False  # this is unreachable, but mypy wants it
+
+
+class IndexingSequenceMixin(Iterable[T_item]):
+    """
+    This mixin allow to extend a list into an object that can be indexed
+    a la mnumpy/pytorch.
+
+    You can index into a IndexingSequenceMixin like a numpy array or torch tensor:
+
+    .. code-block:: python
+        da[0]  # index by position
+        da[0:5:2]  # index by slice
+        da[[0, 2, 3]]  # index by list of indices
+        da[True, False, True, True, ...]  # index by boolean mask
+
+    You can delete items from a DocumentArray like a Python List
```

Comment:
So no deletion with torch-like indexing? Is there a technical limitation that requires this? I can see that being surprising/confusing

Reply:
Well I copy paste your implementation and your docstring :sweat_smile:

Reply:
https://github.com/docarray/docarray/blob/76ecf8c62768b134538e660ce36bb2bb8ec5e28f/docarray/array/array/array.py#L115

Reply:
@JohannesMessner what should I add for this "torch-like" indexing

Reply:
```suggestion
```